File: demo-b/renderer/src/filename_builder.py
"""Filename builder utility for export standardization.

Generates deterministic filenames based on preset usage and modification status.
"""

import json
import os
import sys
from pathlib import Path
from typing import Optional, Dict, Any


def load_preset(preset_id: Optional[str]) -> Optional[Dict[str, Any]]:
    """Load export preset by ID.

    Args:
        preset_id: Preset ID (e.g., 'A2_Paper_v1')

    Returns:
        Preset object or None if not found
    """
    if not preset_id:
        return None

    # Try multiple possible paths
    possible_paths = [
        Path(__file__).parent.parent.parent.parent / 'config' / 'export_presets' / f'{preset_id}.json',
        Path('/app/config/export_presets') / f'{preset_id}.json',
        Path(Path.cwd()) / 'config' / 'export_presets' / f'{preset_id}.json'
    ]

    for preset_path in possible_paths:
        if preset_path.exists():
            try:
                with open(preset_path, 'r', encoding='utf-8') as f:
                    return json.load(f)
            except Exception as e:
                print(f"[FilenameBuilder] Failed to parse preset {preset_id} from {preset_path}: {e}", file=sys.stderr)

    return None


def field_differs(preset: Dict[str, Any], field_path: str, request_value: Any) -> bool:
    """Check if a field value differs from preset default.

    Args:
        preset: Preset object
        field_path: Dot-separated path (e.g., 'render.dpi')
        request_value: Value from request

    Returns:
        True if value differs
    """
    if not preset:
        return False

    parts = field_path.split('.')
    preset_value = preset
    for part in parts:
        if preset_value and isinstance(preset_value, dict):
            preset_value = preset_value.get(part)
        else:
            return True  # Field doesn't exist in preset, consider it different

    # Compare values (handle numbers and strings)
    if isinstance(request_value, (int, float)) and isinstance(preset_value, (int, float)):
        return abs(request_value - preset_value) > 0.001  # Allow small floating point differences

    return request_value != preset_value


def is_preset_modified(preset: Dict[str, Any], request_params: Dict[str, Any]) -> bool:
    """Check if preset was modified (any non-locked field differs from preset).

    Args:
        preset: Preset object
        request_params: Request parameters

    Returns:
        True if preset was modified
    """
    if not preset or 'constraints' not in preset:
        return False

    constraints = preset['constraints']

    # Check DPI if not locked
    if not constraints.get('dpi_locked', False) and 'dpi' in request_params:
        dpi_value = int(request_params['dpi']) if isinstance(request_params['dpi'], (int, str)) else None
        if dpi_value is not None and field_differs(preset, 'render.dpi', dpi_value):
            return True

    # Check format if not locked
    if not constraints.get('format_locked', False) and 'format' in request_params:
        if field_differs(preset, 'render.format', request_params['format']):
            return True

    # Check theme if not locked
    if not constraints.get('theme_locked', False) and 'theme' in request_params:
        if field_differs(preset, 'theme', request_params['theme']):
            return True

    # Check paper size if not locked (width_mm, height_mm)
    if 'width_mm' in request_params or 'height_mm' in request_params:
        preset_paper = preset.get('paper', {})
        preset_width = preset_paper.get('width_mm')
        preset_height = preset_paper.get('height_mm')

        request_width = float(request_params.get('width_mm', 0)) if 'width_mm' in request_params else None
        request_height = float(request_params.get('height_mm', 0)) if 'height_mm' in request_params else None

        if preset_width is not None and request_width is not None:
            if abs(request_width - preset_width) > 0.1:
                return True
        if preset_height is not None and request_height is not None:
            if abs(request_height - preset_height) > 0.1:
                return True

    # Check layers if not locked
    if not constraints.get('layers_locked', False) and 'layers' in request_params:
        request_layers = request_params['layers']
        preset_layers = preset.get('layers', {})

        if isinstance(request_layers, dict) and isinstance(preset_layers, dict):
            for key, value in request_layers.items():
                if preset_layers.get(key) != value:
                    return True

    return False


def sanitize_filename(s: str) -> str:
    """Sanitize string for use in filename (ASCII-safe, filesystem-safe).

    Args:
        s: String to sanitize

    Returns:
        Sanitized string
    """
    if not s:
        return ''

    import re
    # Replace non-ASCII and problematic characters
    sanitized = re.sub(r'[^a-zA-Z0-9_-]', '_', s)
    sanitized = re.sub(r'_{2,}', '_', sanitized)  # Replace multiple underscores with single
    sanitized = sanitized.strip('_')  # Remove leading/trailing underscores
    return sanitized


def build_export_filename(
    bbox_preset: str,
    dpi: int,
    format_type: str,
    preset_id: Optional[str] = None,
    request_params: Optional[Dict[str, Any]] = None
) -> str:
    """Build export filename according to standardization rules.

    Args:
        bbox_preset: Bbox preset name (e.g., 'stockholm_core')
        dpi: DPI value
        format_type: File format ('png', 'pdf', etc.)
        preset_id: Optional export preset ID (e.g., 'A2_Paper_v1')
        request_params: Additional request parameters for modification check

    Returns:
        Generated filename
    """
    # Sanitize bbox_preset
    safe_bbox = sanitize_filename(bbox_preset or 'custom')

    # Sanitize format (remove leading dot if present)
    safe_format = sanitize_filename(format_type or 'png').lstrip('.')

    # Format DPI
    dpi_str = f"{int(dpi) if dpi else 150}dpi"

    # If no preset_id, use custom format
    if not preset_id:
        return f"{safe_bbox}__custom__{dpi_str}.{safe_format}"

    # Load preset to check if modified
    preset = load_preset(preset_id)

    if not preset:
        # Preset not found, but preset_id was provided - treat as custom
        print(f"[FilenameBuilder] Preset {preset_id} not found, using custom format", file=sys.stderr)
        return f"{safe_bbox}__custom__{dpi_str}.{safe_format}"

    # Check if preset was modified
    modified = is_preset_modified(preset, request_params or {})

    # Sanitize preset ID
    safe_preset_id = sanitize_filename(preset_id)

    if modified:
        return f"{safe_bbox}__{safe_preset_id}_modified__{dpi_str}.{safe_format}"
    else:
        return f"{safe_bbox}__{safe_preset_id}__{dpi_str}.{safe_format}"

File: demo-a/exporter/src/filenameBuilder.js
/**
 * Filename builder utility for export standardization.
 *
 * Generates deterministic filenames based on preset usage and modification status.
 */

const path = require('path');
const fs = require('fs');

/**
 * Load export preset by ID.
 * @param {string} presetId - Preset ID (e.g., 'A2_Paper_v1')
 * @returns {object|null} Preset object or null if not found
 */
function loadPreset(presetId) {
  if (!presetId) return null;

  // Try multiple possible paths
  const possiblePaths = [
    path.join(__dirname, '../../../config/export_presets', `${presetId}.json`), // From demo-a/exporter/src to repo root
    path.join('/app/config/export_presets', `${presetId}.json`), // Docker container path
    path.join(process.cwd(), 'config/export_presets', `${presetId}.json`), // From repo root
    path.join(process.cwd(), '../../config/export_presets', `${presetId}.json`) // From demo-a/exporter
  ];

  for (const presetPath of possiblePaths) {
    if (fs.existsSync(presetPath)) {
      try {
        const content = fs.readFileSync(presetPath, 'utf8');
        return JSON.parse(content);
      } catch (e) {
        console.warn(`[FilenameBuilder] Failed to parse preset ${presetId} from ${presetPath}:`, e.message);
      }
    }
  }

  return null;
}

/**
 * Check if a field value differs from preset default.
 * @param {object} preset - Preset object
 * @param {string} fieldPath - Dot-separated path (e.g., 'render.dpi')
 * @param {any} requestValue - Value from request
 * @returns {boolean} True if value differs
 */
function fieldDiffers(preset, fieldPath, requestValue) {
  if (!preset) return false;

  const parts = fieldPath.split('.');
  let presetValue = preset;
  for (const part of parts) {
    if (presetValue && typeof presetValue === 'object') {
      presetValue = presetValue[part];
    } else {
      return true; // Field doesn't exist in preset, consider it different
    }
  }

  // Compare values (handle numbers and strings)
  if (typeof requestValue === 'number' && typeof presetValue === 'number') {
    return Math.abs(requestValue - presetValue) > 0.001; // Allow small floating point differences
  }

  return requestValue !== presetValue;
}

/**
 * Check if preset was modified (any non-locked field differs from preset).
 * @param {object} preset - Preset object
 * @param {object} requestParams - Request parameters
 * @returns {boolean} True if preset was modified
 */
function isPresetModified(preset, requestParams) {
  if (!preset || !preset.constraints) return false;

  const constraints = preset.constraints;

  // Check DPI if not locked
  if (!constraints.dpi_locked && requestParams.dpi !== undefined) {
    if (fieldDiffers(preset, 'render.dpi', parseInt(requestParams.dpi))) {
      return true;
    }
  }

  // Check format if not locked
  if (!constraints.format_locked && requestParams.format !== undefined) {
    if (fieldDiffers(preset, 'render.format', requestParams.format)) {
      return true;
    }
  }

  // Check theme if not locked
  if (!constraints.theme_locked && requestParams.theme !== undefined) {
    if (fieldDiffers(preset, 'theme', requestParams.theme)) {
      return true;
    }
  }

  // Check paper size if not locked (width_mm, height_mm)
  if (requestParams.width_mm !== undefined || requestParams.height_mm !== undefined) {
    const presetWidth = preset.paper?.width_mm;
    const presetHeight = preset.paper?.height_mm;
    const requestWidth = parseFloat(requestParams.width_mm);
    const requestHeight = parseFloat(requestParams.height_mm);

    if (presetWidth !== undefined && Math.abs(requestWidth - presetWidth) > 0.1) {
      return true;
    }
    if (presetHeight !== undefined && Math.abs(requestHeight - presetHeight) > 0.1) {
      return true;
    }
  }

  // Check layers if not locked
  if (!constraints.layers_locked && requestParams.layers !== undefined) {
    // Parse layers if it's a string
    let requestLayers = requestParams.layers;
    if (typeof requestLayers === 'string') {
      try {
        requestLayers = JSON.parse(requestLayers);
      } catch (e) {
        // Ignore parse errors
      }
    }

    if (requestLayers && typeof requestLayers === 'object' && preset.layers) {
      for (const [key, value] of Object.entries(requestLayers)) {
        if (preset.layers[key] !== value) {
          return true;
        }
      }
    }
  }

  return false;
}

/**
 * Sanitize string for use in filename (ASCII-safe, filesystem-safe).
 * @param {string} str - String to sanitize
 * @returns {string} Sanitized string
 */
function sanitizeFilename(str) {
  if (!str) return '';

  // Replace non-ASCII and problematic characters
  return str
    .replace(/[^a-zA-Z0-9_-]/g, '_')
    .replace(/_{2,}/g, '_') // Replace multiple underscores with single
    .replace(/^_+|_+$/g, ''); // Remove leading/trailing underscores
}

/**
 * Build export filename according to standardization rules.
 *
 * @param {object} options - Filename options
 * @param {string} options.bbox_preset - Bbox preset name (e.g., 'stockholm_core')
 * @param {string} [options.preset_id] - Export preset ID (e.g., 'A2_Paper_v1')
 * @param {number|string} options.dpi - DPI value
 * @param {string} options.format - File format ('png', 'pdf', etc.)
 * @param {object} [options.requestParams] - Additional request parameters for modification check
 * @returns {string} Generated filename
 */
function buildExportFilename({ bbox_preset, preset_id, dpi, format, requestParams = {} }) {
  // Sanitize bbox_preset
  const safeBbox = sanitizeFilename(bbox_preset || 'custom');

  // Sanitize format (remove leading dot if present)
  const safeFormat = sanitizeFilename(format || 'png').replace(/^\./, '');

  // Format DPI
  const dpiStr = `${parseInt(dpi) || 150}dpi`;

  // If no preset_id, use custom format
  if (!preset_id) {
    return `${safeBbox}__custom__${dpiStr}.${safeFormat}`;
  }

  // Load preset to check if modified
  const preset = loadPreset(preset_id);

  if (!preset) {
    // Preset not found, but preset_id was provided - treat as custom
    console.warn(`[FilenameBuilder] Preset ${preset_id} not found, using custom format`);
    return `${safeBbox}__custom__${dpiStr}.${safeFormat}`;
  }

  // Check if preset was modified
  const modified = isPresetModified(preset, requestParams);

  // Sanitize preset ID
  const safePresetId = sanitizeFilename(preset_id);

  if (modified) {
    return `${safeBbox}__${safePresetId}_modified__${dpiStr}.${safeFormat}`;
  } else {
    return `${safeBbox}__${safePresetId}__${dpiStr}.${safeFormat}`;
  }
}

module.exports = {
  buildExportFilename,
  loadPreset,
  isPresetModified,
  sanitizeFilename
};

